### PR TITLE
many: add supports for keeping and finding assertions with different format iterations

### PR DESCRIPTION
**Pull request by pedronis**

### CODE REVIEW

**Comment by pedronis** [comment]
based #2148

File: asserts/fsbackstore.go [hunk 1]
```diff
@@ -51,7 +52,7 @@ func OpenFSBackstore(path string) (Backstore, error) {
 }
 
 // guarantees that result assertion is of the expected type (both in the AssertionType and go type sense)
-func (fsbs *filesystemBackstore) readAssertion(assertType *AssertionType, diskPrimaryPath string) (Assertion, error) {
+func (fsbs *filesystemBackstore) readAssertion1(assertType *AssertionType, diskPrimaryPath string) (Assertion, error) {
```
Comment (niemeyer) [inline asserts/fsbackstore.go]: I assume the 1 here is a left over from a prior design?  Can probably be moved back now.

File: asserts/snap_asserts.go [hunk 2]
```diff
@@ -121,46 +121,51 @@ func assembleSnapDeclaration(assert assertionBase) (Assertion, error) {
...
-		for iface, rule := range plugs {
-			plugRule, err := compilePlugRule(iface, rule)
-			if err != nil {
-				return nil, err
-			}
-			plugRules[iface] = plugRule
+	var slotRules map[string]*SlotRule
+
+	// guard evolvable complex parts
+	if assert.SupportedFormat() {
```
Comment (niemeyer) [inline asserts/snap_asserts.go]: Feels like a slightly fragile approach. If we don't understand the format, we should probably assume we also don't understand why we don't understand the format. Perhaps we might control this on the general part of the parsing by using the format header itself to guard the information that changed. Or perhaps prevent the assertion from being read altogether?

File: asserts/findwildcard.go [hunk 3]
```diff
@@ -78,6 +92,9 @@ func findWildcardDescend(top, current string, descendantWithWildcard []string, f
 	if err != nil {
 		return err
 	}
+	if len(descendantWithWildcard) == 1 {
```
Comment (niemeyer) [inline asserts/findwildcard.go]: Thanks for pulling off all those tweaks this late.

File: asserts/findwildcard.go [hunk 4]
```diff
@@ -27,42 +27,56 @@ import (
 )
 
 /*
-findWildcard invokes foundCb for each regular file matching:
+findWildcard invokes foundCb once for each parent directory of regulars files matching:
```
Comment (emgee) [inline asserts/findwildcard.go]: s/regulars/regular

File: overlord/assertstate/assertmgr_test.go [hunk 5]
```diff
@@ -502,6 +502,8 @@ func (s *assertMgrSuite) TestValidateSnapCrossCheckFail(c *C) {
 }
 
 func (s *assertMgrSuite) TestValidateSnapSnapDeclIsTooNewFirsInstall(c *C) {
```
Comment (emgee) [inline overlord/assertstate/assertmgr_test.go]: s/Firs/First

Comment (emgee) [inline overlord/assertstate/assertmgr_test.go]: s/SnapSnap/Snap

Comment (pedronis) [inline overlord/assertstate/assertmgr_test.go]: the first Snap comes from ValidateSnap the name of the tested function

**Pull request merged**